### PR TITLE
Github Action to automatically sync content

### PR DESCRIPTION
**Pull request by stp-ip**
Content is synced from coredns/coredns and checked before being
committed into the same repository so master should always reflect the
correct state.

It is using a scheduled github action for now as this removes any github
authentication coordination and simplifies the general action.

This prepares further automation for Netlify.

Being run every 12 minutes starting on the 4th minute.

Note to reviewer:
We currently push changes even so corecheck fails. We should analyse on, which it fails and maybe fix or rewrite them to not use "corefile" as syntax to remove checking.

### CODE REVIEW

**Comment by stp-ip** [comment]
We could move this to be run on a dispatch event that would be run during release or similar. Basically sending a POST request with a personal github token.

Wanted to be simple first.

File: .github/workflows/sync.coredns.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+          repository: coredns/coredns
+          path: .coredns
+      -
+        name: Setup Go Env
+        run: |
+          echo "::set-env name=GOPATH::$(go env GOPATH)"
+          echo "::add-path::$(go env GOPATH)/bin"
+        shell: bash
+      -
+        name: Deps
```
Comment (miekg) [inline .github/workflows/sync.coredns.yml]: why do need we need to build coredns?

File: Makefile [hunk 2]
```diff
@@ -1,25 +1,15 @@
-C:=~/src/github.com/coredns/coredns
-
-all:
-	hugo -d /var/www/coredns.io
-
-.PHONY: clean
-clean:
-	rm -rf /var/www/coredns.io/*
-
-.PHONY: test
```
Comment (miekg) [inline Makefile]: this might be still useful?

File: .github/workflows/sync.coredns.yml (hunk 1 above)
Comment (stp-ip) [reply .github/workflows/sync.coredns.yml]: The CoreDNS binary is needed by corecheck afaik.
As we already have the source code building seemed like the easiest solution especially considering we are checking the corefiles within the documentation of that specific commit of coredns anyway

File: Makefile (hunk 2 above)
Comment (stp-ip) [reply Makefile]: Happy to keep something along the lines of `make server`. Will push that.

Comment (stp-ip) [reply Makefile]: Kept it as `make run` see latest commit.

File: .github/workflows/sync.coredns.yml (hunk 1 above)
Comment (miekg) [reply .github/workflows/sync.coredns.yml]: hmm, you're right; seems to be heavy weight to just pushing to netifly; we better leave this out and do it off line (for now). Also if coredns doesn't compile we can't deploy (just mentioning this)

**Comment by miekg** [comment]
Note that make sync *always* changes the date on the plugin's metadata; we may want to change that as well in the python code; use the current month or something, not the current time otherwise we will always be pushing changes out

File: .github/workflows/sync.coredns.yml (hunk 1 above)
Comment (stp-ip) [reply .github/workflows/sync.coredns.yml]: Deploy on netlify is separate from this.
I'll push a new version, where the corefile check is done in a separate action. That way we have an idea on issues even so we don't cancel a sync etc.

**Comment by stp-ip** [comment]
Will check the time update.

**Comment by stp-ip** [comment]
* Split up corefile test and sync.
* Replaced syncing python code with Go (tried with an AST solution, but simplified it back)
  * Syncing now takes into account content changes and only overwrites, if necessary.
  * Syncing now uses a toml lib to reduce manual string constructs
* Fixed merge conflicts.

**Comment by stp-ip** [comment]
Would love another review @miekg

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns.io] Github Act..." ]
>Would love another review @miekg

oh yeah, sorry. On the TODO. Think it lgtm, and we can just go forward.

Do we have a domain name pointing to it?

**Comment by stp-ip** [comment]
On the netlify side it's currently netlify.coredns.io
Once we feel comfortable with both the auto sync and netlify auto deploy we can switch www.coredns.io or about.coredns.io it. (Or use the apex as we currently do)

We could then redirect coredns.io to the new subdomain.

**Comment by stp-ip** [comment]
* coredns.netlify.com

**Comment by stp-ip** [comment]
Doing a follow up PR to fix the go script. Seems to update time even so it shouldn't.

**Pull request merged**